core/src/main/java/io/grpc/Context.java

```diff
@@ -0,0 +1,226 @@
+package io.grpc;
+
+import com.google.common.base.Preconditions;
+
+import java.io.Closeable;
+import java.util.concurrent.Callable;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.ScheduledExecutorService;
+
+import javax.annotation.Nullable;
+
+
+/**
+ * --Add warning about using this too liberally--
+ * --Add warning about using this with containers that unload Classloaders--
+ *
+ *
+ */
+// Discussion point. Using java.io.Closeable here so we can inherit AutoCloseable where its available.
+public final class Context implements Closeable {
+
+  /**
+   * The logical root context which is {@link #current()} if no other context is bound. This context
+   * is already closed and will not retain any listeners.
+   */
+  public static final Context ROOT = new Context();
+  static {
+    ROOT.close();
```

Comment:
Let's say you have:

```java
Context someOpenContext = Context.current();
Context ctx1 = someOpenContext.withValue(key1, value);
Context ctx2 = ctx1.withValue(key2, value);
Context ctx3 = ctx2.withValue(key3, value);
Context ctx4 = ctx3.withListener(listener3);
```

Because ctx1-4 each can be closed separately and because there is a listener in ctx4, there _must_ be four or five listeners registered. One listener is the one provided via `withListener()`. But you also need a listener internally from ctx4 to ctx3 (this one _could_ be removed), ctx3 to ctx2, ctx2 to ctx1, and ctx1 to someOpenContext.

Otherwise notifications will be incorrect with things like this:
```java
ctx1.withListener(listener1);
ctx2.close();
```

If only certain Contexts were closeable, then you wouldn't need a full chain. Links like ctx4 to ctx3 are unnecessary since ctx3 is not closeable; ctx4 (or even listener3) could link directly to someOpenContext.

With every `Context` being closeable, fluent-style breaks down a bit:
```java
Context newContext = Context.current().withValue(key, specialValue).withValue(key2, value2);
newContext.attach();
try {
  // work
} finally {
  // This doesn't clean up the Context containing specialValue!
  // If it had any listener registered it would leak until its parent was closed.
  newContext.detachAndClose();
}
```